### PR TITLE
Editor Onboarding: Enable tooltip for 50% of users

### PR DESCRIPTION
**Pull request by AmandaRiu**
Enables the editor onboarding tooltip for 50% of users. The process to determine if the user can view the editor onboarding tooltip has also been updated to do the following:
1. Get the ID of the user. If the user logged into the app via the site address option for a self-hosted site (i.e - not a WPcom account), then we use the anonymous `userId` generated by Tracks and saved in `SharedPrefs`. 
2. Append "can_view_editor_onboarding" to the `userId` as a seed to keep the calculated hashed result unique to this feature.
3. Hash the `userId+seed` using SHA-256
4. Convert the hash byte array to a `BigInteger`
5. Divide the `BigInteger` by 100 to get the _remainder_, which will be a value between 0 and 100. 
6. If the result from step 5 is greater than or equal to 50, then the user can view the onboarding tooltip.

### To test:
The test steps in this section only verify that the onboarding tooltipis displayed properly when the editor is opened for the very first time after a fresh install to ensure nothing was broken in the process of adding the changes included in this PR. To actually verify the changes in this PR, I'd recommend logging the results from the new `getUserId()` and `convertUserIdToHash()` methods and watching the console output while working through these test steps. The `getUserId()` method should NOT return `null` or an empty or `0` string. The `convertUserIdToHash()` method should return a value between 0 and 100. 

Repeat the following test steps using a **WPcom account**, and then logging into a **self-hosted site** (non-Jetpacked) **using the site address option**. 

1. Uninstall WordPress from the test device and then install using the build from this PR or branch.
2. Log into the app.
3. Create a new blog post. Verify the **Tap to add content** tooltip appears directly above the inserter button (`+`).
4. Close the editor to return to My Site page. 
5. Tap the button to create another new blog post. Verify the tooltip does not appear. 

## Regression Notes
1. Potential unintended areas of impact: NONE.

2. What I did to test those areas of impact (or what existing automated tests I relied on): N/A

3. What automated tests I added (or what prevented me from doing so): N/A

PR submission checklist:

- [x] I have completed the Regression Notes.
- [x] I have considered adding accessibility improvements for my changes.
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

### CODE REVIEW

File: WordPress/src/main/java/org/wordpress/android/ui/posts/EditPostActivity.java [hunk 1]
```diff
@@ -2316,9 +2320,51 @@ private GutenbergPropsBuilder getGutenbergPropsBuilder() {
...
+     * Uses cryptography to hash a userId plus a seed containing the feature flag, converts
+     * the result into a BigInteger, then divides that value by 100 and returns the remainder.
+     * The value returned will be used to determine if a user should see the editor onboarding
+     * tooltip.
+     *
+     * @param userId A unique ID for the logged in user.
+     * @param seed The feature string to keep hashed value results between features unique.
+     * @return An int between 1 and 100
+     */
+    private int convertUserIdToHash(String userId, String seed) {
```
Comment (dcalhoun) [inline WordPress/src/main/java/org/wordpress/android/ui/posts/EditPostActivity.java:2356]: I speak from limited knowledge, but would `salt` be a [more appropriate](https://en.wikipedia.org/wiki/Salt_(cryptography)) name than `seed`?

File: WordPress/src/main/java/org/wordpress/android/ui/posts/EditPostActivity.java [hunk 2]
```diff
@@ -3216,7 +3262,6 @@ public void onVideoPressInfoRequested(final String videoId) {
 
     @Override
     public void onEditorFragmentInitialized() {
-        boolean shouldFinishInit = true;
```
Comment (dcalhoun) [inline WordPress/src/main/java/org/wordpress/android/ui/posts/EditPostActivity.java:3219]: This change appears unrelated to the rest of this work. Was this intentional? What does it address?

File: WordPress/src/main/java/org/wordpress/android/ui/posts/EditPostActivity.java [hunk 3]
```diff
@@ -2296,7 +2300,7 @@ private GutenbergPropsBuilder getGutenbergPropsBuilder() {
         boolean isFreeWPCom = mSite.isWPCom() && SiteUtils.onFreePlan(mSite);
         boolean isWPComSite = mSite.isWPCom() || mSite.isWPComAtomic();
 
-        boolean enableEditorOnboarding = canViewEditorOnboarding() && !AppPrefs.hasLaunchedGutenbergEditor();
+        boolean enableEditorOnboarding = !AppPrefs.hasLaunchedGutenbergEditor() && canViewEditorOnboarding();
```
Comment (dcalhoun) [inline WordPress/src/main/java/org/wordpress/android/ui/posts/EditPostActivity.java:2303]: Was this change merely a personal preference? This has no impact on the outcome, correct? I do not have a preference either way, just confirming.

**Review by dcalhoun — APPROVED** [review]
LGTM. I tested the changes on a Samsung Galaxy S20 both with and without a WPCom account. The expected values logged for `getUserId` and `convertUserIdToHash`. The tooltip also displayed as expected on the first editor session due to the enablement for all debug/alpha builds. 🎉 

I left a few comments, but they should not necessarily block this work. Feel free to ignore them as you see fit. 

Thank you for completing this work!

File: WordPress/src/main/java/org/wordpress/android/ui/posts/EditPostActivity.java (hunk 2 above)
Comment (mchowning) [reply WordPress/src/main/java/org/wordpress/android/ui/posts/EditPostActivity.java:3219]: Looks like this variable was just unused, and [the commit message](https://github.com/wordpress-mobile/WordPress-Android/pull/14897/commits/ae550ae70e6b59638c624e873f53f2fc40b17763) confirms that is why she removed it.

File: WordPress/src/main/java/org/wordpress/android/ui/posts/EditPostActivity.java (hunk 1 above)
Comment (mchowning) [reply WordPress/src/main/java/org/wordpress/android/ui/posts/EditPostActivity.java:2356]: Somewhat related question, instead of passing this as a parameter, would it be simpler to instead just define the seed/salt inside of this method? The fact that it is being passed in makes me think that this is something we're expecting to change which, if I'm understanding correctly, we actually don't want or need the seed/salt to change, even if we were to use this with multiple different features. 

This is definitely **not** a blocker for the PR.

File: WordPress/src/main/java/org/wordpress/android/ui/posts/EditPostActivity.java [hunk 4]
```diff
@@ -2316,9 +2320,51 @@ private GutenbergPropsBuilder getGutenbergPropsBuilder() {
...
+     * be removed when the functionality is opened to all users.
+     */
+    private boolean canViewEditorOnboarding() {
+        // Get the userId for the logged in user. If not WPcom, use the anonymous userId
+        // generated by tracks.
+        long userAccountId = mAccountStore.getAccount().getUserId();
+        String userId = String.valueOf(userAccountId);
+        if (userAccountId == 0) {
+            SharedPreferences preferences = PreferenceManager.getDefaultSharedPreferences(getApplicationContext());
+            userId = preferences.getString("nosara_tracks_anon_id", null);
```
Comment (mchowning) [inline WordPress/src/main/java/org/wordpress/android/ui/posts/EditPostActivity.java:2335]: Instead of hardcoding this string, would we be better off reusing it from [`AnalyticsTrackerNosaroa.TRACKS_ANON_ID`](https://github.com/wordpress-mobile/WordPress-Android/blob/7348943876467cd0ff3e7e9fa572e62c5c7abc07/libs/analytics/WordPressAnalytics/src/main/java/org/wordpress/android/analytics/AnalyticsTrackerNosara.java#L19)? We would need to expose either the `TRACKS_ANON_ID` constant or the `getAnonIdPrefKey`, but that seems preferable to repeating the string here. 

Or, perhaps we should find a way to access [`Tracker::getAnonId`](https://github.com/wordpress-mobile/WordPress-Android/blob/a3e0e8833a3a09f9be3035619388bc84a817eaf3/libs/analytics/WordPressAnalytics/src/main/java/org/wordpress/android/analytics/Tracker.java#L53-L59) here, instead? That will probably require a bit of work to expose it through the relevant wrappers, but I think I lean towards that option myself.

File: WordPress/src/main/java/org/wordpress/android/ui/posts/EditPostActivity.java (hunk 1 above)
Comment (mchowning) [reply WordPress/src/main/java/org/wordpress/android/ui/posts/EditPostActivity.java:2356]: Do we need a salt here? I thought that a salt was only necessary to increase how easy it would be to reverse the hashing algorithm (i.e., it makes the cryptography more secure), but in this case we don't care if it is easy to decode it, so I'm wondering if we could simplify this by dropping the salt. Or maybe we're using teh salt because adding it gives us a better distribution?

I do not feel strongly about this, so it's not a blocker as far as I'm concerned.

Comment (dcalhoun) [reply WordPress/src/main/java/org/wordpress/android/ui/posts/EditPostActivity.java:2356]: > Do we need a salt here? 

@mchowning it was my understanding that including the salt is the foundation of the idea to avoid including the same set of users for user-ID-based calculations. Without a unique salt, which we are using the experiment name, we would likely be right back to including the same set of users for any calculation across various experiments/rollouts. 

> Somewhat related question, instead of passing this as a parameter, would it be simpler to instead just define the seed/salt inside of this method? The fact that it is being passed in makes me think that this is something we're expecting to change which, if I'm understanding correctly, we actually don't want or need the seed/salt to change, even if we were to use this with multiple different features.

My understanding is that the salt would have to change if we used this approach for a different experiment. If it did not, then all the hashes would be the same between the two experiments, meaning the same set of users would be included for both experiments.

Comment (mchowning) [reply WordPress/src/main/java/org/wordpress/android/ui/posts/EditPostActivity.java:2356]: You're right. Thanks for explaining. That also explains why it makes sense to include it in the function signature. Might be good to expand on that point in the function's comment to make sure someone else doesn't make the same mistake I made.

Comment (AmandaRiu) [reply WordPress/src/main/java/org/wordpress/android/ui/posts/EditPostActivity.java:2356]: Thanks @mchowning and @dcalhoun for the feedback - I really appreciate it. Since the hashing part is a temporary measure for a project on a tight timeline I'm good leaving this as-is, but will definitely keep all this in mind for the future. 👍

File: WordPress/src/main/java/org/wordpress/android/ui/posts/EditPostActivity.java (hunk 4 above)
Comment (AmandaRiu) [reply WordPress/src/main/java/org/wordpress/android/ui/posts/EditPostActivity.java:2335]: Thanks for the review @mchowning! I thought the same thing as well, but since this hashing mechanism is a temporary solution and will only be in place until the editor onboarding project is rolled out to all users, I didn't want to make a bunch of changes in the more permanent supporting code. I'm good leaving this string as-is for this implementation. 👍
